optuna/integration/dask.py

```diff
@@ -0,0 +1,738 @@
+import asyncio
+import datetime
+import inspect
+from typing import Any
+from typing import Callable
+from typing import Dict
+from typing import Generator
+from typing import List
+from typing import Optional
+import uuid
+
+import optuna
+from optuna._experimental import experimental
+from optuna._imports import try_import
+from optuna.distributions import BaseDistribution
+from optuna.distributions import distribution_to_json
+from optuna.distributions import json_to_distribution
+from optuna.storages import BaseStorage
+from optuna.study import StudyDirection
+from optuna.study import StudySummary
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+with try_import() as _imports:
+    import distributed
+    from distributed.utils import thread_state
+    from distributed.worker import get_client
+
+# Serialization utilities
+
+
+def _serialize_datetime(dt: datetime.datetime) -> dict:
```

Comment:
[question] These serialize/deserialize functions refer to internal implementation of `frozen trial`, `studysummary` and `studydirection`. If their implementations are changed, we need to update these serialize/deserialize functions. Do you have any idea to solve this problem?

Reply:
That's a great point @keisuke-umezawa. Do you know how stable these objects are? I don't have a good sense for how often the internals of these objects are refactored. There may be some precedence for these types of utilities as they seem quite similar to the [existing `distribution_to_json`/`json_to_distribution` functions](https://github.com/optuna/optuna/blob/84cff7ee4191cc7266d49e2f2ffa85e40999d558/optuna/distributions.py#L464-L501)

> Was a bit curious about this as well. Is it the case that pickle won't do?

Unfortunately not. For security reasons, Dask's scheduler doesn't run pickle (specifically it doesn't deserialize with pickle) since that can cause arbitrary bytecode to be executed. Instead the scheduler uses [`msgpack`](https://github.com/msgpack/msgpack-python) for serializing/deserializing objects, which is safer than pickle from a security perspective, but also more limited in the types of objects it can handle.

This is why these additional serialization functions were introduced here. So we can extract the relevant state off of objects (e.g. `TrialState`) in a `msgpack` serializable form that Dask's scheduler can handle.